### PR TITLE
Avoid division by zero when denominator is Nullable

### LINKED ISSUE #22621: Null in decimal datatype, division by zero.
**Describe the bug**
There is a lot of similar reports (https://github.com/ClickHouse/ClickHouse/issues/7509) but they mostly related to https://github.com/ClickHouse/ClickHouse/issues/12587 and If/case statements.

**How to reproduce**
Clickhouse version 21.5
```
SELECT
    1 / CAST(NULL, 'Nullable(UInt32)') AS int,
    1 / CAST(NULL, 'Nullable(UInt32)') AS float

Query id: 8f68936f-3ff9-4637-bc63-c19fc71a58ed

┌──int─┬─float─┐
│ ᴺᵁᴸᴸ │  ᴺᵁᴸᴸ │
└──────┴───────┘

SELECT 1 / CAST(NULL, 'Nullable(Decimal(7, 2))')

Query id: ed010d6a-4924-4af3-a200-701deeb9d2f4


0 rows in set. Elapsed: 0.002 sec.

Received exception from server (version 21.5.1):
Code: 153. DB::Exception: Received from localhost:9000. DB::Exception: Division by zero: While processing 1 / CAST(NULL, 'Nullable(Decimal(7, 2))').
```

### PR DESCRIPTION
**Pull request by Avogar**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Avoid error "Division by zero" when  denominator is Nullable in functions `divide`, `intDiv` and `modulo`. Closes #22621

Detailed description / Documentation draft:
When we have Nullable(Something) column, in rows where value is Null, nested column can contain random values (often it's default value - 0) and when we perform binary operation (like divide, intDiv or modulo) we perform it on nested column and only then make the result Nullable according to null masks of arguments. So, we get division by zero error because we cannot divide by zero. This PR fixes this problem by adding special implementation for the case when we have Nullable(Something) column as a denominator.

### CODE REVIEW

**Comment by qoega** [comment]
probably pr-improvement? It always worked that way before?

**Comment by Avogar** [comment]
> probably pr-improvement? It always worked that way before?

It always worked that way, pr-improvement is more appropriate, thanks.

File: src/Functions/FunctionBinaryArithmetic.h [hunk 1]
```diff
@@ -367,18 +390,62 @@ struct DecimalBinaryOperation
...
+                    c[i] = applyScaledDiv<is_decimal_a>(
+                        unwrap<op_case, OpCase::LeftConstant>(a, i), unwrap<op_case, OpCase::RightConstant>(b, i), scale_a);
             return;
         }
 
-        for (size_t i = 0; i < size; ++i)
-            c[i] = apply(
-                unwrap<op_case, OpCase::LeftConstant>(a, i),
-                unwrap<op_case, OpCase::RightConstant>(b, i));
+        if (right_nullmap)
```
Comment (CurtizJ) [inline src/Functions/FunctionBinaryArithmetic.h:421]: Looks like too much boilerplate code. Maybe we can somehow generalize it?

**Comment by CurtizJ** [comment]
@Avogar, please solve conflicts and we will merge this PR.

**Review by CurtizJ — APPROVED** [review]

**Pull request merged**